fastrand.go

```diff
@@ -53,16 +53,21 @@ func (r *randReader) Read(b []byte) (int, error) {
 	// counter so that concurrent callers also end up with unique values.
 	counter := atomic.AddUint64(&r.counter, 1)
 	counterExtra := atomic.LoadUint64(&r.counterExtra)
-	// Update the second 64 bits of the counter if the first 64 bits are close
-	// to wrapping around. It is possible that the second 64 bits of the counter
-	// is updates multiple times by several concurrent threads. This wastes part
-	// of the counter space (up to 2^63 items each time), however the overall
-	// space is large enough (2^128) that wasting some every reset does not make
-	// it any more likely that the caller exhaust the whole possible search
-	// space.
-	if counter > 1<<63 {
+
+	// Update the second 64 bits of the counter if needed. Because there are
+	// multiple threads, it is possible for the counter to have incremented
+	// beyond '1<<63' by the time coutnerExtra is incremented. So we perform an
+	// additional increment at math.MaxUint64 to make sure that potential
+	// overlaps can never result in the same counter being used twice.
+	//
+	// Note that there could still be a problem if 2^63 threads all manage to
+	// increment 'counter' before counterExtra is incremented by the first
+	// thread. This has been deemend to be sufficiently unlikely.
```

Comment:
The code looks good but the comment doesn't explain it super well. Granted, it's not an easy thing to explain concisely. Here's my stab at it:
```
// Increment counterExtra when counter is close to overflowing. We cannot wait
// until counter == math.MaxUint64 to increment counterExtra, because another
// goroutine could call Read, overflowing counter to 0 before the first
// goroutine increments counterExtra. The second goroutine would then be
// reusing the counter pair (0, 0). Instead, we increment at 1<<63 so that
// there is little risk of an overflow.
//
// There is still a potential overlap near 1<<63, though, because another
// goroutine could see counter == 1<<63+1 before the first goroutine
// increments counterExtra. The counter pair (1<<63+1, 1) would then be
// reused. To prevent this, we also increment at math.MaxUint64. This means
// that in order for an overlap to occur, 1<<63 goroutine would need to
// increment counter before the first goroutine increments counterExtra.
```
Also, I think the branch could be:
```go
if counter == 1<<63 || counter == math.MaxUint64 {
```
Since we're already assuming that `counter` won't be incremented `1<<63` times during a single goroutine